crates/bevy_ecs/src/entity/allocator.rs

```diff
@@ -0,0 +1,771 @@
+        // The capacity is only one bit, and if other bits needed to be dropped, `leading` would have been greater
+        let slice_index = full_idnex & !Self::capacity_of_chunk(chunk_index);
+
+        (chunk_index, slice_index)
+    }
+
+    /// Gets the entity at the index within this chunk.
+    ///
+    /// # Safety
+    ///
+    /// [`Self::set`] must have been called on this index before.
+    #[inline]
+    unsafe fn get(&self, index: u32) -> Entity {
+        // SAFETY: caller ensure we are init.
+        let head = unsafe { self.ptr().debug_checked_unwrap() };
+        // SAFETY: caller ensures we are in bounds (because `set` must be in bounds)
+        let target = unsafe { &*head.add(index as usize) };
+
+        target.get_entity()
+    }
+
+    /// Gets a slice of indices.
+    ///
+    /// # Safety
+    ///
+    /// [`Self::set`] must have been called on these indices before.
+    #[inline]
+    unsafe fn get_slice(&self, index: u32, ideal_len: u32, index_of_self: u32) -> &[Slot] {
+        let cap = Self::capacity_of_chunk(index_of_self);
+        let after_index_slice_len = cap - index;
+        let len = after_index_slice_len.min(ideal_len) as usize;
+
+        // SAFETY: caller ensure we are init.
+        let head = unsafe { self.ptr().debug_checked_unwrap() };
+
+        // SAFETY: The chunk was allocated via a `Vec` and the index is within the capacity.
+        unsafe { core::slice::from_raw_parts(head, len) }
+    }
+
+    /// Sets this entity at this index.
+    ///
+    /// # Safety
+    ///
+    /// This must not be called concurrently.
+    /// Index must be in bounds.
+    /// Access does not conflict with another [`Self::get`].
+    #[inline]
+    unsafe fn set(&self, index: u32, entity: Entity, index_of_self: u32) {
+        let head = self.ptr().unwrap_or_else(|| self.init(index_of_self));
+        // SAFETY: caller ensures it is in bounds and we are not fighting with other `set` calls or `get` calls.
+        // A race condition is therefore impossible.
+        let target = unsafe { &*head.add(index as usize) };
+
+        target.set_entity(entity);
+    }
+
+    /// Initializes the chunk to be valid, returning the pointer.
+    ///
+    /// # Safety
+    ///
+    /// This must not be called concurrently.
+    #[cold]
+    unsafe fn init(&self, index: u32) -> *mut Slot {
+        let cap = Self::capacity_of_chunk(index);
+        let mut buff = ManuallyDrop::new(Vec::new());
+        buff.reserve_exact(cap as usize);
+        let ptr = buff.as_mut_ptr();
+        self.first.store(ptr, Ordering::Relaxed);
+        ptr
+    }
+
+    /// Frees memory
+    ///
+    /// # Safety
+    ///
+    /// This must not be called concurrently.
+    unsafe fn dealloc(&self, index: u32) {
+        if let Some(to_drop) = self.ptr() {
+            let cap = Self::capacity_of_chunk(index) as usize;
+            // SAFETY: This was created in [`Self::init`] from a standard Vec.
+            unsafe {
```

Comment:
Is this sound? We only ever reserved the vec, we didn't resize the vec. That means that technically a valid `Slot` does not exist at each index in the Vec, but we are about to drop the Vec which expects to drop all those (assumed valid) slots. We could be dropping uninitialized memory. I know we don't define a Drop impl (this needs to be called out in the safety comment BTW), but I'm not sure if Rust also has some expectations for dropping AtomicU32?

Similarly, when we `set` the slots, is it sound to assign to uninitialized memory there? Because we are treating uninitialized memory as a valid `Slot`.